File: typescript-sdk/tests/test_session_cleanup.ts
/**
 * Connected-socket cleanup: Session.create/close and the sessionList/vault*
 * free functions must always destroy their transport when the wrapped RPC
 * fails — an open net.Socket is an active libuv handle that keeps the
 * process alive and leaks an fd per retry.
 *
 * Observation: the MockDaemon counts open server-side connections; a
 * client-side transport.close() (socket destroy) drops the count back to
 * zero.
 */
import { test, describe, before, after } from "node:test";
import assert from "node:assert/strict";
import { Session, sessionList, vaultGrant, LoomRPCError } from "../src/index.js";
import { MockDaemon } from "./helpers/mock_daemon.js";

async function waitForNoOpenConnections(daemon: MockDaemon, timeoutMs = 2000): Promise<void> {
  const start = Date.now();
  while (daemon.openConnectionCount > 0) {
    if (Date.now() - start > timeoutMs) {
      throw new Error(
        `timed out: daemon still sees ${daemon.openConnectionCount} open connection(s) — ` +
          "a transport was leaked",
      );
    }
    await new Promise((r) => setTimeout(r, 10));
  }
}

describe("socket cleanup on RPC failure", () => {
  let daemon: MockDaemon;

  before(async () => {
    daemon = new MockDaemon();
    await daemon.start();
  });

  after(async () => {
    await daemon.stop();
  });

  test("Session.create failure closes the connected transport", async () => {
    daemon.registerHandler("session.create", () => {
      throw new Error("unknown_profile");
    });
    try {
      await assert.rejects(
        () => Session.create({ socketPath: daemon.socketPath, token: daemon.token }),
        LoomRPCError,
      );
      await waitForNoOpenConnections(daemon);
    } finally {
      // Restore the default handler for the rest of the suite.
      daemon.registerHandler("session.create", () => ({
        session_id: "01TEST" + "1".repeat(20),
        status: "active",
        created_at_ms: Date.now(),
      }));
    }
  });

  test("Session.close failure still closes the transport", async () => {
    daemon.registerHandler("session.close", () => {
      throw new Error("boom");
    });
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    await assert.rejects(() => s.close(), LoomRPCError);
    await waitForNoOpenConnections(daemon);
  });

  test("sessionList failure closes the connected transport", async () => {
    daemon.registerHandler("session.list", () => {
      throw new Error("boom");
    });
    await assert.rejects(
      () => sessionList({ socketPath: daemon.socketPath, token: daemon.token }),
      LoomRPCError,
    );
    await waitForNoOpenConnections(daemon);
  });

  test("vaultGrant failure closes the connected transport", async () => {
    // No vault.grant handler registered → method_not_found error envelope.
    await assert.rejects(
      () =>
        vaultGrant("01TEST", "https://example.com", ["read"], 60, "label", {
          socketPath: daemon.socketPath,
          token: daemon.token,
        }),
      (err: unknown) => {
        assert.ok(err instanceof LoomRPCError);
        assert.strictEqual((err as LoomRPCError).code, "method_not_found");
        return true;
      },
    );
    await waitForNoOpenConnections(daemon);
  });
});

File: python-sdk/tests/test_session_cleanup.py
"""
Connected-socket cleanup: Session.create / Session.close (sync and async)
must always release their transport when the wrapped RPC fails, instead of
leaking the connected Unix socket.
"""

from __future__ import annotations

import pytest

import loom
from loom._async_transport import AsyncLoomTransport
from loom._errors import LoomRPCError
from loom._transport import LoomTransport


def _fail(_params: dict) -> dict:
    raise ValueError("boom")  # MockDaemon converts this to an internal_error envelope


def test_session_create_failure_closes_transport(daemon, monkeypatch):
    created: list[LoomTransport] = []

    class RecordingTransport(LoomTransport):
        def __init__(self, *args, **kwargs) -> None:
            super().__init__(*args, **kwargs)
            created.append(self)

    monkeypatch.setattr(loom, "LoomTransport", RecordingTransport)
    daemon.register_handler("session.create", _fail)
    with pytest.raises(LoomRPCError):
        loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token)
    assert len(created) == 1
    # socket.close() detaches the fd (fileno() == -1) — the socket was released.
    assert created[0]._sock.fileno() == -1


def test_session_close_failure_still_closes_transport(daemon):
    daemon.register_handler("session.close", _fail)
    s = loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token)
    with pytest.raises(LoomRPCError):
        s.close()
    assert s._transport._sock.fileno() == -1


async def test_async_session_create_failure_closes_transport(daemon, monkeypatch):
    created: list[AsyncLoomTransport] = []

    class RecordingTransport(AsyncLoomTransport):
        def __init__(self, *args, **kwargs) -> None:
            super().__init__(*args, **kwargs)
            created.append(self)

    monkeypatch.setattr(loom, "AsyncLoomTransport", RecordingTransport)
    daemon.register_handler("session.create", _fail)
    with pytest.raises(LoomRPCError):
        await loom.AsyncSession.create(socket_path=str(daemon.socket_path), token=daemon.token)
    assert len(created) == 1
    assert created[0]._closed is True


async def test_async_session_close_failure_still_closes_transport(daemon):
    daemon.register_handler("session.close", _fail)
    s = await loom.AsyncSession.create(socket_path=str(daemon.socket_path), token=daemon.token)
    with pytest.raises(LoomRPCError):
        await s.close()
    assert s._transport._closed is True
